contributors/design-proposals/watchable-discovery.md

```diff
@@ -0,0 +1,231 @@
+all resources. The name of group objects is the API group name. The
+Kind is cluster-scoped.
+
+```golang
+// +genclient=true
+// +nonNamespaced=true
+
+// Group describes an API group with all its versions and resources.
+type Group struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Status of the group.
+	Status GroupStatus `json:"spec,omitempty" protobuf:"bytes,2,opt,name=status"`
+}
+
+// GroupStatus describes the API group with all its versions and resources.
+type GroupStatus struct {
+	// versions are the versions supported in this group, order by preference. The first one is the preferred
+	// version for the group.
+	Versions []GroupVersion `json:"versions" protobuf:"bytes,1,rep,name=versions"`
+	// a map of client CIDR to server address that is serving this group.
+	// This is to help clients reach servers in the most network-efficient way possible.
+	// Clients can use the appropriate server address as per the CIDR that they match.
+	// In case of multiple matches, clients should use the longest matching CIDR.
+	// The server returns only those CIDRs that it thinks that the client can match.
+	// For example: the master will return an internal IP CIDR only, if the client reaches the server using an internal IP.
+	// Server looks at X-Forwarded-For header or X-Real-Ip header or request.RemoteAddr (in that order) to get the client IP.
+	ServerAddressByClientCIDRs []metav1.ServerAddressByClientCIDR `json:"serverAddressByClientCIDRs" protobuf:"bytes,2,rep,name=serverAddressByClientCIDRs"`
+}
+
+// GroupVersion describes one version of an API group, including all resources available in this version.
+type GroupVersion struct {
+	// the version name
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// resources contains the name of the resources and if they are namespaced.
+	Resources []metav1.APIResource `json:"resources" protobuf:"bytes,2,rep,name=resources"`
+}
+
+// GroupList is a list of API groups.
+type GroupList struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// List of group.
+	Items []Group `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+```
+
+## Behavior
+
+Each API server in the cluster serves its own `discovery.k8s.io` API
+group. By default, this is backed by normal etcd storage, i.e. the
+semantics of the `ResourceVersion` is the one of the underlaying etcd
+storage backend.
+
+Each API server has its own `DiscoveryUpdateController` that's
+regularly reconciles the storage state of the discovery API group with
+its internal registered/announce API group registry. Moreover, when
+a new API groups is installed or removed in its HTTP handler, the
+corresponding API group object is updated synchronously.
+
+### Aggegration
+
+The aggregator watches the discovery API group of each of its proxied
```

Comment:
The aggregator is the only one who decides which GVR is visible. It's the place to do the aggregation. It also does that now with the polling btw: it selectively forwards discovery calls to the right API server, hiding those which are not to be exposed. The only difference is that it also takes care of the resources, not only GV.